Fix TestEgress flaky failure in ipv6-only-e2e-test

wenqiq opened the pull request:
Fix TestEgress flaky failure in jenkins-ipv6-only-e2e

The test failure is affected by the same IPv6 address in
`testEgressClientIP` test case.
The deleting Pod process in `defer deletePodWrapper(t, data, tt.fakeServer)`
deleted the EgressIP which used by `testEgressNodeFailure` when concurrently
running.
It only happens to only-ipv6-e2e because IPv6 test uses the same IP for
multiple tests while IPv4 uses different IPs. It has enough time to delete
Pod between different test cases in dual-stack cluster e2e.

This patch fixes it by changing the IPv6 addresses used in different
testEgress cases.

Resolves #2718

Signed-off-by: Wenqi Qiu <wenqiq@vmware.com>

wenqiq commented:
/test-ipv6-only-e2e

wenqiq commented:
It seems the failure is affected by the same IPv6 address in testEgressClientIP test case. When deleting Pod in `defer deletePodWrapper(t, data, tt.fakeServer)`.The same IPv6 address was deleted.
Antrea-controller log:
```
I0905 15:30:40.440899       1 controller.go:570] Processing Egress egress-l2lbk UPDATE event with selector ({&LabelSelector{MatchLabels:map[string]string{foo: bar,},MatchExpressions:[]LabelSelectorRequirement{},} nil []})
I0905 15:30:40.441054       1 controller.go:515] Updating existing EgressGroup egress-l2lbk with 0 Pods on 0 Nodes
I0905 15:30:40.441067       1 controller.go:456] Finished syncing Egress egress-l2lbk. (19.417µs)
I0905 15:30:40.441004       1 controller.go:358] "Updated ExternalIPPool status" ExternalIPPool="pool-ttc9p"
I0905 15:30:40.619234       1 controller.go:570] Processing Egress egress-l2lbk UPDATE event with selector ({&LabelSelector{MatchLabels:map[string]string{foo: bar,},MatchExpressions:[]LabelSelectorRequirement{},} nil []})
I0905 15:30:40.619294       1 controller.go:515] Updating existing EgressGroup egress-l2lbk with 0 Pods on 0 Nodes
I0905 15:30:40.619307       1 controller.go:456] Finished syncing Egress egress-l2lbk. (22.648µs)
I0905 15:30:41.145516       1 controller.go:180] Processing Pod antrea-test/remotepodipv6-cluster UPDATE event, labels: map[antrea-e2e:remotepodipv6-cluster app:busybox]
I0905 15:30:41.306250       1 controller.go:180] Processing Pod antrea-test/localpodipv6-cluster UPDATE event, labels: map[antrea-e2e:localpodipv6-cluster app:busybox]
I0905 15:30:41.313188       1 controller.go:180] Processing Pod antrea-test/eth-ipv6 UPDATE event, labels: map[antrea-e2e:eth-ipv6 app:agnhost]
I0905 15:30:42.148335       1 controller.go:180] Processing Pod antrea-test/remotepodipv6-cluster UPDATE event, labels: map[antrea-e2e:remotepodipv6-cluster app:busybox]
```

So c.localIPDetector.IsLocalIP check result is false then the sync process updated the Egress status to ''.`	if err := c.updateEgressStatus(egress, c.localIPDetector.IsLocalIP(egress.Spec.EgressIP)); err != nil {
		return fmt.Errorf("update Egress %s status error: %v", egressName, err)
	}`

Antrea-agent log:
```
I0905 15:30:40.435237       1 egress_controller.go:542] "Updating Egress status" Egress="egress-l2lbk" oldNode="" newNode="antrea-ipv6-2-0"
I0905 15:30:40.437269       1 egress_controller.go:265] "Detected ExternalIPPool event" ExternalIPPool="pool-ttc9p" enqueueEgressNum=1
I0905 15:30:40.440659       1 egress_controller.go:554] "Updated Egress status" Egress="egress-l2lbk"
I0905 15:30:40.440731       1 ip_assigner_linux.go:122] "The IP is already assigned" ip="2021::aaa2"
I0905 15:30:40.440821       1 ip_assigner_linux.go:146] "Sent NDP neighbor advertisement" ip="2021::aaa2"
I0905 15:30:40.440858       1 egress_controller.go:554] "Updated Egress status" Egress="egress-l2lbk"
I0905 15:30:40.605034       1 egress_controller.go:249] Detected Egress IP address 2021::aaa2 deleted from this Node
I0905 15:30:40.605137       1 ip_assigner_linux.go:122] "The IP is already assigned" ip="2021::aaa2"
I0905 15:30:40.608005       1 ip_assigner_linux.go:146] "Sent NDP neighbor advertisement" ip="2021::aaa2"
I0905 15:30:40.611689       1 iptables.go:225] "Deleted a rule" rule=[-m comment --comment Antrea: SNAT Pod to external packets ! -o antrea-gw0 -m mark --mark 0x00000001/0x000000ff -j SNAT --to 2021::aaa2] table="nat" chain="ANTREA-POSTROUTING" protocol="IPv6"
I0905 15:30:40.614414       1 egress_controller.go:542] "Updating Egress status" Egress="egress-l2lbk" oldNode="antrea-ipv6-2-0" newNode=""
I0905 15:30:40.621568       1 egress_controller.go:554] "Updated Egress status" Egress="egress-l2lbk"
I0905 15:30:40.621604       1 ip_assigner_linux.go:122] "The IP is already assigned" ip="2021::aaa2"
I0905 15:30:40.621667       1 ip_assigner_linux.go:146] "Sent NDP neighbor advertisement" ip="2021::aaa2"
I0905 15:30:40.621683       1 egress_controller.go:554] "Updated Egress status" Egress="egress-l2lbk"
```

It only happens in only-ipv6-e2e because it has enough time to delete Pod between different test case in testEgressClientIP.

Change the IPv6 address in different testEgress case is a solution for flaky fail in ipv6-only-e2e.

wenqiq commented:
/test-ipv6-only-e2e

/test-ipv6-e2e
/test-ipv6-only-conformance

/test-ipv6-conformance
/test-ipv6-networkpolicy
/test-ipv6-only-networkpolicy

tnqn commented on test/e2e/egress_test.go:
```diff
@@ -717,21 +719,18 @@ func (data *TestData) createEgress(t *testing.T, generateName string, matchExpre
 
 func (data *TestData) waitForEgressRealized(egress *v1alpha2.Egress) (*v1alpha2.Egress, error) {
-	err := wait.PollImmediate(200*time.Millisecond, 5*time.Second, func() (done bool, err error) {
+	err := wait.PollImmediate(200*time.Millisecond, waitNodeFailureTimeout, func() (done bool, err error) {
```
It doesn't look suitable to use waitNodeFailureTimeout for waiting for egress realized

wenqiq replied:
Done
Has been updated to `waitEgressRealizedTimeout`.

tnqn commented on test/e2e/egress_test.go:
```diff
@@ -600,10 +602,10 @@ func testEgressMigration(t *testing.T, data *TestData, triggerFunc, revertFunc f
 		},
 	}
-	externalIPPoolTwoNodes := data.createExternalIPPool(t, "pool-", *ipRange, matchExpressions, nil)
+	externalIPPoolTwoNodes := data.createExternalIPPool(t, "two-node-pool-", *ipRange, matchExpressions, nil)
```
two-node-pool sounds confusing. Maybe pool-with-two-nodes

wenqiq replied:
Done

tnqn reviewed (commented):
Code LGTM, two minor comments

tnqn commented:
Please briefly explain how it fixes the failure in the commit message

wenqiq commented:
/test-ipv6-only-e2e

tnqn commented:
> It only happens in only-ipv6-e2e because it has enough time to
delete Pod between different test cases in testEgressClientIP

I don't see difference in time between ipv4 and ipv6 tests. Isn't it because ipv6 test uses same IP for multiple tests while IPv4 uses different IPs?

> Change the IPv6 address in different testEgress cases
is a solution for flaky fail in ipv6-only-e2e.

**Changing** the IPv6 address in different testEgress cases is a solution for flaky **failure** in ipv6-only-e2e.

wenqiq commented:
> > It only happens in only-ipv6-e2e because it has enough time to
> > delete Pod between different test cases in testEgressClientIP
> 
> I don't see difference in time between ipv4 and ipv6 tests. Isn't it because ipv6 test uses same IP for multiple tests while IPv4 uses different IPs?

Updated. 
It explains why it didn't happen to dual-stack cluster test.

> > Change the IPv6 address in different testEgress cases
> > is a solution for flaky fail in ipv6-only-e2e.
> 
> **Changing** the IPv6 address in different testEgress cases is a solution for flaky **failure** in ipv6-only-e2e.

Updated.

wenqiq commented:
/test-ipv6-only-e2e

tnqn commented:
```
Fix TestEgress flaky fail in jenkins-ipv6-only-e2e 

The test failure is affected by the same IPv6 address in
`testEgressClientIP` test case.
When deleting Pod in `defer deletePodWrapper(t, data, tt.fakeServer)`.
The same IPv6 address in `testEgressNodeFailure` was deleted.
It only happens to only-ipv6-e2e because IPv6 test uses the same IP
for multiple tests while IPv4 uses different IPs.

Changing the IPv6 address in different testEgress cases
is a solution for flaky failure in ipv6-only-e2e.
```

There are still some grammer and format issue in the commit message, could you follow https://github.com/antrea-io/antrea/blob/main/CONTRIBUTING.md#getting-reviewers to improve it?
The issues I can see:
1. Fix TestEgress flaky **failure**  in jenkins-ipv6-only-e2e
2. The first paragraph is weird, it starts a newline starting from "When deleting ..." without an empty line
3. The lines are not wrapped at same length, some have just 50+ chars while some have 70+
4. The last paragraph should just say "This patch fixes it by changing the IPv6 addresses used in different testEgress cases" to be clear. "XXXX is a solution" sounds like a suggestion when debugging the issue, while in commit message we should just say what we actually do.

tnqn commented:
```
The test failure is affected by the same IPv6 address in
`testEgressClientIP` test case.
This patch fixes it by changing the IPv6 addresses used in different
testEgress cases.
```

Again, there is format issue in the commit message. Either don't start a new paragraph for the second sentence or insert an empty line between paragraphs.

wenqiq commented:
Thanks. Good advice!
I will read this article more carefully. https://chris.beams.io/posts/git-commit/
`What may be a hassle at first soon becomes habit, and eventually a source of pride and productivity for all involved.`

wenqiq commented:
/test-ipv6-only-all

tnqn commented:
s/probabilistically/probabilistic
probabilistically is adv.

tnqn reviewed (approved):
LGTM

tnqn commented:
/test-e2e
/test-ipv6-e2e
/test-ipv6-only-e2e
/skip-conformance
/skip-networkpolicy

tnqn commented:
/test-integration

wenqiq commented:
/test-integratio

wenqiq commented:
/test-integration

wenqiq commented:
> @wenqiq could you backport this to releases that have this issue?

Yes, done, only release-1.3 has this issue.
